core/domain/details.go

```diff
@@ -0,0 +1,173 @@
+package domain
+
+import (
+	"fmt"
+
+	"github.com/anyproto/any-store/anyenc"
+	"github.com/gogo/protobuf/types"
+)
+
+// Detail is Key-Value pair
+type Detail struct {
+	Key   RelationKey
+	Value Value
+}
+
+type Details = GenericMap[RelationKey]
+
+func NewDetails() *Details {
+	return &GenericMap[RelationKey]{data: make(map[RelationKey]Value, 20)}
+}
+
+func NewDetailsFromProto(st *types.Struct) *Details {
+	data := make(map[RelationKey]Value, len(st.GetFields()))
+	d := &GenericMap[RelationKey]{data: data}
+	for k, v := range st.GetFields() {
+		d.SetProtoValue(RelationKey(k), v)
+	}
+	return d
+}
+
+func NewDetailsFromMap(details map[RelationKey]Value) *Details {
+	return &GenericMap[RelationKey]{
+		data: details,
+	}
+}
+
+func NewDetailsWithSize(size int) *Details {
+	return &GenericMap[RelationKey]{data: make(map[RelationKey]Value, size)}
+}
+
+func NewDetailsFromAnyEnc(v *anyenc.Value) (*Details, error) {
+	obj, err := v.Object()
+	if err != nil {
+		return nil, fmt.Errorf("is object: %w", err)
+	}
+	res := NewDetailsWithSize(obj.Len())
+	var visitErr error
+	obj.Visit(func(k []byte, v *anyenc.Value) {
+		if visitErr != nil {
+			return
+		}
+		// key is copied
+		err := setValueFromAnyEnc(res, RelationKey(k), v)
+		if err != nil {
+			visitErr = err
+		}
+	})
+	return res, visitErr
+}
+
+func setValueFromAnyEnc(d *Details, key RelationKey, val *anyenc.Value) error {
+	switch val.Type() {
+	case anyenc.TypeNumber:
+		v, err := val.Float64()
+		if err != nil {
+			return fmt.Errorf("number: %w", err)
+		}
+		d.SetFloat64(key, v)
+		return nil
+
+	case anyenc.TypeString:
+		v, err := val.StringBytes()
+		if err != nil {
+			return fmt.Errorf("string: %w", err)
+		}
+		d.SetString(key, string(v))
+		return nil
+	case anyenc.TypeTrue:
+		d.SetBool(key, true)
+		return nil
+	case anyenc.TypeFalse:
+		d.SetBool(key, false)
+		return nil
+	case anyenc.TypeArray:
+		arrVals, err := val.Array()
+		if err != nil {
+			return fmt.Errorf("array: %w", err)
+		}
+		// Assume string as default type
+		if len(arrVals) == 0 {
+			d.SetStringList(key, nil)
+			return nil
+		}
+
+		firstVal := arrVals[0]
+		if firstVal.Type() == anyenc.TypeString {
+			res := make([]string, 0, len(arrVals))
+			for _, arrVal := range arrVals {
+				v, err := arrVal.StringBytes()
+				if err != nil {
+					return fmt.Errorf("array item: string: %w", err)
+				}
+				res = append(res, string(v))
+			}
+			d.SetStringList(key, res)
+			return nil
+		} else if firstVal.Type() == anyenc.TypeNumber {
+			res := make([]float64, 0, len(arrVals))
+			for _, arrVal := range arrVals {
+				v, err := arrVal.Float64()
+				if err != nil {
+					return fmt.Errorf("array item: number: %w", err)
+				}
+				res = append(res, v)
+			}
+			d.SetFloat64List(key, res)
+			return nil
+		} else {
+			return fmt.Errorf("unsupported array type %s", firstVal.Type())
+		}
+	}
+	d.Set(key, Null())
+	return nil
+}
+
+// StructDiff returns pb struct which contains:
```

Comment:
You can clarify by linking to the concept of [symmetric difference](https://en.wikipedia.org/wiki/Symmetric_difference) for better understanding.